File: src/leetCodeProblems/PriorityQueue/MinimumNoOfRefuelingStops871.java
package leetCodeProblems.PriorityQueue;

/**
 * https://leetcode.com/problems/minimum-number-of-refueling-stops/submissions/
 */

import java.util.Comparator;
import java.util.HashSet;
import java.util.PriorityQueue;

public class MinimumNoOfRefuelingStops871 {

    static class GasStation {
        int position;
        int fuel;

        GasStation(int position, int fuel) {
            this.position = position;
            this.fuel = fuel;
        }
    }

    static class PriorityQueueCompare implements Comparator<GasStation> {

        public int compare(GasStation station1, GasStation station2) {

            return station2.fuel - station1.fuel;
        }
    }

    private int isMaxAlreadyTraversed(HashSet<Integer> alreadyTraversedStations, int currentStationPosition) {

        for (Integer station: alreadyTraversedStations) {
            if (station >= currentStationPosition) {
                return station;
            }
        }

        return -1;
    }

    private void rebuiltQueue(PriorityQueue<GasStation> gasStationQueue, PriorityQueue<GasStation> tempQueue) {

        for (GasStation gasStation: gasStationQueue) {
            tempQueue.add(gasStation);
        }
    }

    public int minRefuelStops(int target, int startFuel, int[][] stations) {

        if (startFuel >= target) {
            return 0;
        }

        if (stations.length == 0) {
            return -1;
        }

        PriorityQueue<GasStation> gasStationQueue =
                new PriorityQueue<GasStation>(stations.length, new PriorityQueueCompare());


        for(int i=0; i < stations.length; i++) {

            GasStation station = new GasStation(stations[i][0], stations[i][1]);

            gasStationQueue.add(station);
        }

        /*System.out.println(gasStationQueue.remove().fuel);
        System.out.println(gasStationQueue.remove().fuel);
        System.out.println(gasStationQueue.remove().fuel);
        System.out.println(gasStationQueue.remove().fuel);*/

        int currentFuel = startFuel;
        int currentTarget = target;
        int currentPosition = 0;

        HashSet<Integer> alreadyTraversedStations = new HashSet<Integer>();

        int ans = 0;
        int stationsTraversed = 0;


        while (currentTarget > currentFuel) {

            if (gasStationQueue.size() == 0) {
                return -1;
            }

            PriorityQueue<GasStation> tempQueue =
                    new PriorityQueue<GasStation>(gasStationQueue.size(), new PriorityQueueCompare());

            rebuiltQueue(gasStationQueue, tempQueue);

            while (tempQueue.size() > 0) {

                GasStation maxStation = tempQueue.remove();

                int neededFuel = maxStation.position - currentPosition;

                System.out.println("maxStation.position =>" + maxStation.position);
                System.out.println("neededFuel => " + neededFuel);
                if (neededFuel <= currentFuel) {

                    int maxAlreadyTraversed = isMaxAlreadyTraversed(alreadyTraversedStations, maxStation.position);

                    if ( maxAlreadyTraversed!= -1) {
                        currentFuel = currentFuel + maxStation.fuel;
                        currentTarget = target - maxAlreadyTraversed;
                        currentPosition = maxAlreadyTraversed;
                    }
                    else {
                        currentFuel = currentFuel - neededFuel + maxStation.fuel;
                        currentTarget = target - maxStation.position;
                        currentPosition = maxStation.position;
                    }

                    System.out.println("currentFuel => " + currentFuel);
                    System.out.println("currentTarget => " + currentTarget);
                    System.out.println("currentPosition => " + currentPosition);
                    gasStationQueue.remove(maxStation);

                    alreadyTraversedStations.add(maxStation.position);

                    ans++;
                    break;
                }

                else {
                    stationsTraversed++;
                }
            }

            if (stationsTraversed == stations.length) {
                return -1;
            }
        }

        if (ans == 0) {
            ans = -1;
        }

        return ans;
    }

    public static void main(String[] args) {

        MinimumNoOfRefuelingStops871 obj = new MinimumNoOfRefuelingStops871();

        //int[][] stations = {{10,60}, {20,30}, {30,30}, {60,40}};
        //int target = 100;
        //int startFuel = 10; // Expected Output = 2

        //int[][] stations = {{2,4}, {3,2}, {5,5}, {16,10}};
        //int target = 25;
        //int startFuel = 10; // Expected Output = 3

        /*int[][] stations = {{10,100}};
        int target = 100;
        int startFuel = 1;*/

        //int[][] stations = new int[0][0];
        //int target = 1;
        //int startFuel = 1;

        //int[][] stations = new int[0][0];
        //int target = 100;
        //int startFuel = 1;


        //int[][] stations = {{25,25}, {50,25}, {75,25}};
        //int target = 100;
        //int startFuel = 25;

        int[][] stations = {{25,30}};
        int target = 100;
        int startFuel = 50;

        System.out.println(obj.minRefuelStops(target, startFuel, stations));

    }
}

File: src/leetCodeProblems/Greedy/MinimumNoOfRefuelingStops871.go
package main

/*
- LeetCode - https://leetcode.com/problems/minimum-number-of-refueling-stops/
- Time - O(N*logN)
- Space - O(N)
*/
import (
	"container/heap"
	"log"
	"sort"
)

type stopsQueue []int

func (h stopsQueue) Len() int {
	return len(h)
}

func (h stopsQueue) Less(i int, j int) bool {
	return h[i] > h[j]
}

func (h stopsQueue) Swap(i int, j int) {
	h[i], h[j] = h[j], h[i]
}

func (h *stopsQueue) Push(a interface{}) {
	*h = append(*h, a.(int))
}

func (h *stopsQueue) Pop() interface{} {
	l := len(*h)
	res := (*h)[l-1]
	*h = (*h)[:l-1]
	return res
}

func minRefuelStops(target int, startFuel int, stations [][]int) int {
	sort.SliceStable(stations, func(i, j int) bool {
		return stations[i][0] < stations[j][0]
	})

	pq := new(stopsQueue)
	out := 0

	stationsLength := len(stations)

	for i := 0; i <= stationsLength; i++ {

		log.Println(pq)
		dist := 0

		if i == stationsLength {
			dist = target
		} else {
			dist = stations[i][0]
		}

		log.Println(startFuel, dist)
		for startFuel < dist {

			if pq.Len() == 0 {
				return -1
			}

			startFuel += heap.Pop(pq).(int)
			out++
		}

		if i < stationsLength {
			heap.Push(pq, stations[i][1])
		}
	}

	return out
}

// func main() {

// 	// stations := [][]int{{10, 60}, {20, 30}, {30, 30}, {60, 40}}
// 	// log.Println(minRefuelStops(100, 10, stations))

// 	stations := [][]int{{25, 30}}
// 	log.Println(minRefuelStops(100, 50, stations))

// }
